File: Parser.java
import java.util.StringTokenizer;
import java.util.HashMap;

class Parser {

    public static void main(String[] args) {
        setupStates();
        runTests();
    }

    public static void runTests() {

        String test1 = "the lazy smelly dog ran EOI"; // Multiple adjectives
        String test2 = "the lazy smelly dog ran slowly EOI"; // Multiple adjectives with adverb
        String test3 = "the lazy dog ate slowly EOI"; // One adjective
        String test4 = "the cat ate EOI"; // No adjectives
        String test5 = "the cat ran slowly EOI"; // No adjectives with adverb

        assert validateSentence(test1);
        assert validateSentence(test2);
        assert validateSentence(test3);
        assert validateSentence(test4);
        assert validateSentence(test5);

        String test6 = "smelly cat ran EOI";
        String test7 = "the cat ate";
        String test8 = "the cat slowly EOI";
        String test9 = "the lazy ate EOI";
        String test10 = "the dog EOI";

        assert !validateSentence(test6);
        assert !validateSentence(test7);
        assert !validateSentence(test8);
        assert !validateSentence(test9);
        assert !validateSentence(test10);

    }

    public static void setupStates() {
        State s4 = new State(null);
        State s3 = new State(new HashMap<>() {
            {
                put("slowly", s4);
                put("noisily", s4);
                put("EOI", s4);
            }
        });

        State s2 = new State(new HashMap<>() {
            {
                put("ate", s3);
                put("ran", s3);
            }
        });

        State s1 = new State(new HashMap<>());
        s1.transitions = new HashMap<>() {
            {
                put("lazy", s1);
                put("smelly", s1);
                put("dog", s2);
                put("cat", s2);
            }
        };

        State s0 = new State(new HashMap<>() {
            {
                put("the", s1);
            }
        });
        startState = s0;
    }

    private static State startState;

    public static boolean validateSentence(String text) {
        StringTokenizer sT = new StringTokenizer(text);
        String token = null;
        State currentState = startState;
        while (sT.hasMoreTokens() && !currentState.eoi) {
            token = sT.nextToken();
            if (currentState.transitions.containsKey(token)) {
                currentState = currentState.transitions.get(token);
            } else
                return false;
        }
        return currentState.eoi;
    }

    private static class State {

        public boolean eoi;
        public HashMap<String, State> transitions;

        public State(HashMap<String, State> transitions) {
            if (transitions == null)
                eoi = true;
            this.transitions = transitions;
        }
    }
}

File: parser.cpp
#include "parser.h"

parser::parser() {
    machine.insert(make_pair(make_pair(0, "the"), 1));
    machine.insert(make_pair(make_pair(1, "lazy"), 1));
    machine.insert(make_pair(make_pair(1, "smelly"), 1));
    machine.insert(make_pair(make_pair(1, "dog"), 2));
    machine.insert(make_pair(make_pair(1, "cat"), 2));
    machine.insert(make_pair(make_pair(2, "ran"), 3));
    machine.insert(make_pair(make_pair(2, "ate"), 3));
    machine.insert(make_pair(make_pair(3, "noisily"), 4));
    machine.insert(make_pair(make_pair(3, "slowly"), 4));
    machine.insert(make_pair(make_pair(3, "EOI"), 5));
    machine.insert(make_pair(make_pair(4, "EOI"), 5));
}

bool parser::check(vector<string> sentence) {
    int state = 0;
    for (int x = 0; x < sentence.size(); x++) {
        pair<int, string> current = make_pair(state, sentence[x]);
        //call find function from map, if it is not found, the sentence isnt valid
        if (machine.find(current) == machine.end()) {
            return false;
        } else {
            state = machine[current];
        }
    }
    //if it reached the end state, it is a valid sentence
    if (state == 5) {
        return true;
    }
    return false;
}
